lang/cs/Org.Apache.REEF.Network/Elastic/Task/CommunicationLayer.cs

```diff
@@ -0,0 +1,304 @@
+        {
+            var id = NodeObserverIdentifier.FromObserver(operatorObserver);
+
+            if (_driverMessageObservers.ContainsKey(id))
+            {
+                throw new IllegalStateException($"Topology for id {id} already added among driver listeners.");
+            }
+
+            _driverMessageObservers.TryAdd(id, operatorObserver);
+        }
+
+        /// <summary>
+        /// Send the communication message to the task whose name is included in the message.
+        /// </summary>
+        /// <param name="destination">The destination node for the message</param>
+        /// <param name="message">The message to send</param>
+        /// <param name="cancellationSource">The token to cancel the operation</param>
+        internal void Send(
+            string destination,
+            ElasticGroupCommunicationMessage message,
+            CancellationTokenSource cancellationSource)
+        {
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+            if (string.IsNullOrEmpty(destination))
+            {
+                throw new ArgumentNullException(nameof(destination));
+            }
+            if (_disposed)
+            {
+                Log.Log(Level.Warning, "Received send message request after disposing: Ignoring.");
+                return;
+            }
+
+            IIdentifier destId = _idFactory.Create(destination);
+            int retry = 0;
+
+            while (!Send(destId, message))
+            {
+                if (retry > _retrySending)
+                {
+                    throw new IllegalStateException($"Unable to send message after retying {retry} times.");
+                }
+                Thread.Sleep(_timeout);
+
+                retry++;
+            }
+        }
+
+        /// <summary>
+        /// Forward the received message to the target <see cref="IOperatorTopologyWithCommunication"/>.
+        /// </summary>
+        /// <param name="remoteMessage">The received message</param>
+        public abstract void OnNext(IRemoteMessage<NsMessage<ElasticGroupCommunicationMessage>> remoteMessage);
+
+        /// <summary>
+        /// Checks if the identifier is registered with the name server.
+        /// Throws exception if the operation fails more than the retry count.
+        /// </summary>
+        /// <param name="identifiers">The identifier to look up</param>
+        /// <param name="cancellationSource">The token to cancel the operation</param>
+        /// <param name="removed">Nodes that got removed during task registration</param>
+        public void WaitForTaskRegistration(
+            IList<string> identifiers,
+            CancellationTokenSource cancellationSource,
+            ConcurrentDictionary<string, byte> removed = null)
+        {
+            if (removed == null)
+            {
+                removed = new ConcurrentDictionary<string, byte>();
```

Comment:
it is probably better to occasionally check `removed` for null than create a new object here

Reply:
Yea. Fixed.